core/modules/replica/SqlApp.cc

```diff
@@ -0,0 +1,171 @@
+    "This application executes the same query against worker databases of"
+    " select workers. Result sets will be reported upon a completion of"
+    " the application.";
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+SqlApp::Ptr SqlApp::create(int argc, char* argv[]) {
+    return Ptr(
+        new SqlApp(argc, argv)
+    );
+}
+
+
+SqlApp::SqlApp(int argc, char* argv[])
+    :   Application(
+            argc, argv,
+            ::description,
+            true    /* injectDatabaseOptions */,
+            true    /* boostProtobufVersionCheck */,
+            true    /* enableServiceProvider */
+        ) {
+
+    // Configure the command line parser
+
+    parser().required(
+        "user",
+        "Worker-side MySQL user account for executing the query.",
+        _user);
+
+    parser().required(
+        "password",
+        "Password for the MySQL account.",
+        _password);
+
+    parser().required(
+        "query",
+        "The query to be executed on all select workers",
+        _query);
+
+    parser().flag(
+        "all-workers",
+        "The flag for selecting all workers regardless of their status (DISABLED or READ-ONLY).",
+        _allWorkers);
+
+    parser().option(
+        "max-rows",
+        "The maximum number of rows to be pulled from result set at workers"
+        " whene processing queries. NOTE: This parameter nothing to do with"
+        " the SQL's 'LIMIT <num-rows>'. It serves as an additional fail safe"
+        " mechanism preventing protocol buffers from being overloaded by huge"
+        " result sets which might be accidentally initiated by users.",
+        _maxRows);
+
+    parser().option(
+        "worker-response-timeout",
+        "Maximum timeout (seconds) to wait before queries would finish."
+        " Setting this timeout to some reasonably low number would prevent the application from"
+        " hanging for a substantial duration of time (which depends on the default Configuration)"
+        " in case if some workers were down. The parameter applies to operations with"
+        " the Replication workers.",
+        _timeoutSec);
+
+    parser().option(
+        "tables-page-size",
+        "The number of rows in the table of replicas (0 means no pages).",
+        _pageSize);
+}
+
+
+int SqlApp::runImpl() {
+
+    // Limit request execution time if such limit was provided
+    if (_timeoutSec != 0) {
+        serviceProvider()->config()->setControllerRequestTimeoutSec(_timeoutSec);
+    }
+
+    atomic<bool> finished{false};
+    auto const job = SqlJob::create(
+        _query,
+        _user,
+        _password,
+        _maxRows,
+        _allWorkers,
+        Controller::create(serviceProvider()),
+        string(),
+        [&finished] (SqlJob::Ptr const& job) {
+            finished = true;
+        }
+    );
+    job->start();
+
+    util::BlockPost blockPost(1000,2000);
+    while (not finished) {
+        blockPost.wait();
+    }
```

Comment:
spin lock :-( 
there's not a better way?

Reply:
Potentially, I could use `cv`. But this may not worth the effort here. The `BlockPost` allows me to print periodic 'heartbeats' while waiting for the completion of the job.

Reply:
to sure if I'm really "suffering" 😄 However, using `cv` would probably make this code more complex. Besides, this (spill-lock) pattern is mostly (if not to say - exclusively) used in the smile command-line applications. Class `SqlApp` is one of those. All classes ending with `App` are the command-line tools. I need s simple (nothin fancy) mechanism for waiting for a completion of an asynchronous operation.